lib/ABI/FunctionType.cpp

```diff
@@ -388,210 +417,447 @@ class ConversionHelper {
+                                                            Binary,
+                                                            TypeDependencies);
+
+                // Drop the merged half of the current type from the list.
+                Result.pop_back();
+
+                WasHandledSpecifically = true;
+              }
+            }
           }
+        }
 
-          Result.pop_back();
-        } else {
-          return std::nullopt;
+        if (!WasHandledSpecifically) {
+          // TODO: It might still be worth it to add this unused register as
+          // an argument to indicate the register ordering.
         }
+      } else {
+        // Ignore this register, it's most like unused.
       }
 
+      bool IsUsed = (CurrentRegisterIterator != Used.end());
       MustUseTheNextOne = MustUseTheNextOne || IsUsed;
     }
 
+    // Ensure all the indices are consistent.
     for (auto Pair : llvm::enumerate(llvm::reverse(Result)))
       Pair.value().Index = Pair.index();
 
     return Result;
   }
 
-  static llvm::SmallVector<model::Argument, 8>
-  convertStackArguments(model::QualifiedType StackArgumentTypes,
-                        size_t IndexOffset) {
+  /// Helper used for converting stack argument struct into
+  /// the c-style representation
+  ///
+  /// \param StackArgumentTypes The qualified type of the relevant part of
+  ///        the stack.
+  /// \param IndexOffset The index of the first argument (indicates the register
+  ///        argument count).
+  ///
+  /// \return An ordered list of arguments.
+  static std::optional<llvm::SmallVector<model::Argument, 8>>
+  tryConvertingStackArguments(model::QualifiedType StackArgumentTypes,
+                              size_t IndexOffset) {
+    // Qualifiers here are not allowed since this must point to a simple struct.
     revng_assert(StackArgumentTypes.Qualifiers.empty());
+
     auto *Unqualified = StackArgumentTypes.UnqualifiedType.get();
-    if (not Unqualified)
-      return {};
+    if (!Unqualified) {
+      // If there is no type, it means that the importer responsible for this
+      // function didn't detect any stack arguments and avoided creating
+      // a new empty type.
+      return llvm::SmallVector<model::Argument, 8>{};
+    }
 
     auto *Pointer = llvm::dyn_cast<model::StructType>(Unqualified);
     revng_assert(Pointer != nullptr,
                  "`RawFunctionType::StackArgumentsType` must be a struct");
     const model::StructType &Types = *Pointer;
 
+    // If the struct is empty, there's no arguments to extract from it.
+    if (Types.Fields.empty()) {
+      revng_assert(Types.size() == 0);
+      return llvm::SmallVector<model::Argument, 8>{};
+    }
+
+    // Compute the alignment of the first argument.
+    auto CurrentAlignment = *abi::naturalAlignment(Types.Fields.begin()->Type,
+                                                   ABI);
+    if ((CurrentAlignment & (CurrentAlignment - 1)) == 0) {
```

Comment:
`isPowerOf2_64`

Reply:
I went with `isZeroOrPowerOfTwo` in `ABI/Support.h`. Closing this.